File: Question2 Programs/MMD.py
import time

# Function to multiply two matrices of doubles
def MatrixMultiplicationDouble(N):
    A = [[1.0 for _ in range(N)] for _ in range(N)]
    B = [[1.0 for _ in range(N)] for _ in range(N)]
    C = [[0.0 for _ in range(N)] for _ in range(N)]

    start_time = time.time()

    for i in range(N):
        for j in range(N):
            for k in range(N):
                C[i][j] += A[i][k] * B[k][j]

    end_time = time.time()
    print(f"Time taken by double multiplication: {end_time - start_time:.5f} seconds")

if __name__ == "__main__":
    N = 1024
    print(f"Matrix Size: {N}")
    MatrixMultiplicationDouble(N)
    print("-----------------------")

File: Question2 Programs/MMD.cpp
#include <iostream>
#include <vector>
#include <ctime>

using namespace std;

// Function to multiply two matrices of doubles
void MatrixMultiplicationDouble(int N) {
    vector<vector<double>> A(N, vector<double>(N, 1.0));
    vector<vector<double>> B(N, vector<double>(N, 1.0));
    vector<vector<double>> C(N, vector<double>(N, 0.0));
    
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    for(int i = 0; i < N; ++i) {
        for(int j = 0; j < N; ++j) {
            for(int k = 0; k < N; ++k) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &end);
    double time_taken = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
    cout << "Time taken by double multiplication: " << time_taken << " seconds" << endl;
}

int main() {
    int N = 64;

    cout << "Matrix Size: " << N << endl;
    MatrixMultiplicationDouble(N);
    cout << "-----------------------" << endl;

    return 0;
}
